Migrate from `log` to `tracing`

mdrokz opened the pull request:
# Changes

- add `tracing-subscriber` to support tracing in `log-test` 
- add `tracing` with attributes feature
- change log mod to tracing with re exports
- replace all log references to tracing

# Todo

- [ ] add instrument macro to relevant call sites
- [ ] mock the macro when `tracing` is not in use

mdrokz commented:
Also im not able to re export the `instrument` proc macro from the `tracing` mod

```
error[E0432]: unresolved import `crate::tracing::instrument`
  --> src/gsym/parser.rs:42:5
   |
42 | use crate::tracing::instrument;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ no `instrument` in `tracing`
```

any ideas on how i can import the `instrument` proc macro for use ?

d-e-s-o reviewed (commented):
Nice work. Left a few comments. Let me know what you think.

d-e-s-o commented:
> any ideas on how i can import the `instrument` proc macro for use ?

For me
```rust
pub(crate) use tracing::instrument;
```
worked. But there seemingly are some clashes when we rename our internal module to `tracing`.

I also think we cannot easily mock a procedural macro. So we'd likely have to fall back to only using `instrument` conditionally:
```rust
#[cfg_attr(feature = "tracing", crate::log::instrument)]
```

E.g., https://github.com/d-e-s-o/blazesym/commit/70d4d84696c1647c7dfe64897c33d640e69c1dd6

mdrokz commented on Cargo.toml:
```diff
@@ -65,17 +65,17 @@ debug = true
 env_logger = "0.10"
 tracing-subscriber = {version = "0.3", default-features = false, features = ["env-filter", "fmt"]}
-test-log = {version = "0.2", features = ["tracing"]}
+test-log = "0.2"
```
> It doesn't seem right to revert this line, correct? I feel like this commit could just be squashed with the previous one that introduced the `tracing` dependency -- it just gets more convoluted if we split the two.

ah yes that makes sense my bad. So you want me to squash the commit right ?

mdrokz commented on src/lib.rs:
```diff
@@ -91,21 +91,38 @@ impl From<u32> for Pid {
+    pub(crate) use tracing::trace;
+    pub(crate) use tracing::warn;
+
+    #[allow(unused)]
+    pub(crate) use tracing::debug_span;
+
+    #[allow(unused)]
+    pub(crate) use tracing::error_span;
+
+    #[allow(unused)]
+    pub(crate) use tracing::info_span;
+
+    #[allow(unused)]
+    pub(crate) use tracing::trace_span;
+
+    #[allow(unused)]
+    pub(crate) use tracing::warn_span;
```
> Let's add those when we need them, perhaps?

that would work too i added them because i saw `#[allow(unused)]` for `log::info` and others so i thought we could have these as well for later

d-e-s-o replied:
It's subjective, but the others (`info!`, ...) are "core" primitives where once we use one we probably want the others eventually. I don't believe I have ever actually used the `*_span` functionality, to be honest, so it may make more sense to add once we do. Once that is the case, though, it's probably fine to add all the variants (`info`, `error`, etc).

mdrokz replied:
> It's subjective, but the others (`info!`, ...) are "core" primitives where once we use one we probably want the others eventually. I don't believe I have ever actually used the `*_span` functionality, to be honest, so it may make more sense to add once we do. Once that is the case, though, it's probably fine to add all the variants (`info_span`, `error_span`, etc).

cool makes sense

mdrokz commented on src/lib.rs:
```diff
@@ -91,21 +91,38 @@ impl From<u32> for Pid {
 
-#[cfg(feature = "log")]
+#[cfg(feature = "tracing")]
 #[macro_use]
-mod log {
+mod tracing {
```
> It's probably fine to rename the feature, but perhaps we could keep the shorter name `log` for the internal module? It's just a generic logging facility, so I feel like the name is not too out of line. But up to you.

actually i thought that as well as that makes the most sense but i didnt know what you guys would want the name to be. i will revert it back to log

mdrokz commented:
> > any ideas on how i can import the `instrument` proc macro for use ?
> 
> For me
> 
> ```rust
> pub(crate) use tracing::instrument;
> ```
> 
> worked. But there seemingly are some clashes when we rename our internal module to `tracing`.
> 
> I also think we cannot easily mock a procedural macro. So we'd likely have to fall back to only using `instrument` conditionally:
> 
> ```rust
> #[cfg_attr(feature = "tracing", crate::log::instrument)]
> ```
> 
> E.g., [d-e-s-o@70d4d84](https://github.com/d-e-s-o/blazesym/commit/70d4d84696c1647c7dfe64897c33d640e69c1dd6)



> > any ideas on how i can import the `instrument` proc macro for use ?
> 
> For me
> 
> ```rust
> pub(crate) use tracing::instrument;
> ```
> 
> worked. But there seemingly are some clashes when we rename our internal module to `tracing`.
> 
> I also think we cannot easily mock a procedural macro. So we'd likely have to fall back to only using `instrument` conditionally:
> 
> ```rust
> #[cfg_attr(feature = "tracing", crate::log::instrument)]
> ```
> 
> E.g., [d-e-s-o@70d4d84](https://github.com/d-e-s-o/blazesym/commit/70d4d84696c1647c7dfe64897c33d640e69c1dd6)

i tried that as well and it didnt work for me probably because the name might be clashing let me try it again after renaming the mod to be `log`

danielocfb commented:
Looks mostly good! I merged the patch with a few changes. For one I removed the spans from C APIs. I am not sure how much value they add, given that C code is unlikely to ever subscribe to anything. I also adjusted the fields included a bit, cleaned up the history, and added a CHANGELOG entry. Please add a sign-off to your commits next time, make sure that CI passes on your pull request, and resolve merge conflicts with top of tree.

Thanks for getting this feature in!

danielocfb commented:
> I also adjusted the fields included a bit

Just to add to that, if you run
> RUST_LOG=trace RUST_LOG_SPAN_EVENTS=full cargo test -- --nocapture

There was a lot of spam with your original version, because a few huge data fields were included in the output. We should be mindful of that. The output is still not perfect and can be polished, but there is less spam.

mdrokz commented:
> Looks mostly good! I merged the patch with a few changes. For one I removed the spans from C APIs. I am not sure how much value they add, given that C code is unlikely to ever subscribe to anything. I also adjusted the fields included a bit, cleaned up the history, and added a CHANGELOG entry. Please add a sign-off to your commits next time, make sure that CI passes on your pull request, and resolve merge conflicts with top of tree.
> 
> Thanks for getting this feature in!

Thanks for your feedback. I was still working on trying to squash the commits and resolve the merge conflicts, could you clarify signing off on commit i have never done that before.

mdrokz commented:
> > I also adjusted the fields included a bit
> 
> Just to add to that, if you run
> 
> > RUST_LOG=trace RUST_LOG_SPAN_EVENTS=full cargo test -- --nocapture
> 
> There was a lot of spam with your original version, because a few huge data fields were included in the output. We should be mindful of that. The output is still not perfect and can be polished, but there is less spam.

Ah my bad about that i missed this i will keep it in mind next time. thanks
